### PR TITLE
stdlib::ssl: add new stdlib::ssl functions

### PR DESCRIPTION
**Pull request by b4ldr**
@kjetilho pointed out out me recently that validate_x509_rsa_key_pair validates
that the certificate is signed by the private key.  It doesn't validate
that the certificate and key are part of the same key pair.  As such i
have created two new functions

 * `stdlib::ssl::validate_key_pair`
   This vaidates that a public pkey and a private pkey are from the same
   pair
 * `stdlib::ssl::extract_pkey`
   This functions extracts the pkey as a pem from an x509 certificate.

As most users will work directly with x509 certificates the expected use
case would be
```puppet
  stdlib::ssl::validate_key_pair(
    file('ssl/private/key.pem'),
    stdlib::ssl::extract_pkey(file('ssl/public/key.pem')),
  )
```

### CODE REVIEW

File: lib/puppet/functions/stdlib/ssl/extract_pkey.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+require 'openssl'
+# @summary
+#   Extract a public key from a certificate
+Puppet::Functions.create_function(:'stdlib::ssl::extract_pkey') do
```
Comment (b4ldr) [inline lib/puppet/functions/stdlib/ssl/extract_pkey.rb:4]: im not sure if good practrice to add addtional nam spacing like this i.e. `stdlib::ssl::...`

File: lib/puppet/functions/stdlib/ssl/extract_pkey.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+  def extract_pkey(certificate)
+    OpenSSL::X509::Certificate.new(certificate).public_key.to_pem
```
Comment (b4ldr) [inline lib/puppet/functions/stdlib/ssl/extract_pkey.rb:6]: have only tested this with RSA and ed22519 certificates and it it only worked with the RSA one.

File: lib/puppet/functions/stdlib/ssl/validate_key_pair.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  dispatch :validate_ssl_key_pair do
+    param 'String[1]', :public_key
+    param 'String[1]', :private_key
+    return_type 'Boolean'
+  end
+  def validate_ssl_key_pair(public_key, private_key)
+    begin
+      private_key = OpenSSL::PKey.read(private_key)
+    rescue OpenSSL::PKey::PKeyError => error
+      raise Puppet::ParseError, "Not a valid Private key: #{error}"
```
Comment (b4ldr) [inline lib/puppet/functions/stdlib/ssl/validate_key_pair.rb:15]: is it best to raise or just return false here?  or just let the original exception bubble up?

File: lib/puppet/functions/stdlib/ssl/extract_pkey.rb (hunk 2 above)
Comment (bastelfreak) [reply lib/puppet/functions/stdlib/ssl/extract_pkey.rb:6]: which puppetserver? maybe an old jruby?

Comment (b4ldr) [reply lib/puppet/functions/stdlib/ssl/extract_pkey.rb:6]: whatever bundler uses but also using my local ruby 2.7.4p191

```shell
$ bundle exec irb                                                                                             [15:37:24]
irb(main):002:0> require 'openssl'
=> true
irb(main):003:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem')
=> #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name O=Internet Widgits Pty Ltd,ST=Some-State,C=AU>, issuer=#<OpenSSL::X509::Name O=Internet Widgits Pty Ltd,ST=Some-State,C=AU>, serial=#<OpenSSL::BN:0x0000564c7a931800>, not_before=2022-06-03 12:48:56 UTC, not_after=2023-06-03 12:48:56 UTC>
irb(main):004:0> c = OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem')
irb(main):006:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem').public_key.to_pem
Traceback (most recent call last):
       16: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/exe/bundle:49:in `block in <top (required)>'
       15: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:24:in `start'
       14: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
       13: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:30:in `dispatch'
       12: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
       11: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
       10: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        9: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:494:in `exec'
        8: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:28:in `run'
        7: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:63:in `kernel_load'
        6: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:63:in `load'
        5: from /usr/bin/irb:23:in `<top (required)>'
        4: from /usr/bin/irb:23:in `load'
        3: from /usr/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):5
        1: from (irb):6:in `rescue in irb_binding'
NoMethodError (undefined method `to_pem' for #<OpenSSL::PKey::PKey:0x0000564c7a625d00>)
Did you mean?  to_enum
irb(main):007:0>  OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem').public_key.class
=> OpenSSL::PKey::PKey
irb(main):008:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/x509.pem').public_key.to_pem
=> "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAypa/mjZw9cL6uh4HuCoS\n0QOnfmtY5GOl5Gk+eolNUR1M2yacO+QO2FbOnGoO1MpV3AFUhocNM9jwZR8BIRpi\n+Spw2JnnlljXg5qGP9356PVMyImPJYevP7/IE5Yqr32ftBLvXsAPZDh/f9d0/RC8\nEiiwmwWuqjO9zwdy8YjKTei9+U8WU2KpsPB9zBr2yoIaId9NVr/k5mcPbHCSpmL3\nnBYCs3PFaE4Wv7X+/SLDX/V9+aLYX5jC4Gpw1nvEzfkiju4gUl5jy0lRmFckDame\nXJdnQ3q08/ujrwrJ6PFd+fMtjjwz2/XdeDt+hYViezg2GjO+h0KIG7xqXHMLhGb/\nJQIDAQAB\n-----END PUBLIC KEY-----\n"
```

but yes i suspect this is dependent on ruby and the version and implementation of the ssl library you have installed

Comment (alexjfisher) [reply lib/puppet/functions/stdlib/ssl/extract_pkey.rb:6]: Make sure you also do some manual testing with a puppetserver (running puppet via jruby).  I've encountered bugs/inconsistencies with the jruby implementation of openssl before.

File: lib/puppet/functions/stdlib/ssl/validate_key_pair.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+require 'openssl'
+# @summary
+#    Validates a PEM-formatted X.509 public_key and private key using
+#    OpenSSL. Verifies that the public_key's matches the supplied key.
+Puppet::Functions.create_function(:'stdlib::ssl::validate_key_pair') do
+  dispatch :validate_ssl_key_pair do
+    param 'String[1]', :public_key
+    param 'String[1]', :private_key
+    return_type 'Boolean'
```
Comment (alexjfisher) [inline lib/puppet/functions/stdlib/ssl/validate_key_pair.rb:9]: Maybe don't return anything.  All the other `validate_` functions fail or return nothing.

File: lib/puppet/functions/stdlib/ssl/validate_key_pair.rb (hunk 3 above)
Comment (b4ldr) [reply lib/puppet/functions/stdlib/ssl/validate_key_pair.rb:15]: added althugh now wonder if this needs to be split into two separate PR's :(

File: lib/puppet/functions/stdlib/ssl/validate_key_pair.rb (hunk 4 above)
Comment (b4ldr) [reply lib/puppet/functions/stdlib/ssl/validate_key_pair.rb:9]: hmm tbh i didn't like using the validate_ name as it makes me think of type validators before we had custom types, but i couldn't think of a better name. Naming aside  i think returning a boolean is a bit nicer; as an contrived example of the top of my head; you could have a pem file with the full chain and want to loop through and make sure at least one match the private key.

**Comment by chelnak** [comment]
There seems to be a lot of changes in this PR. 

Given that you (@alexjfisher @bastelfreak) have both been active here, what are you thoughts? What can @b4ldr do to move this forward?

**Comment by bastelfreak** [comment]
@chelnak I did say as soon as the CI pass I can give it another review, but mostly this looks good already.

**Comment by chelnak** [comment]
@bastelfreak thanks. There is a bigger issue with the spec tests that we are going to look in to.. but I'll take a look at the validation failures tomorrow.

**Comment by b4ldr** [comment]
Sorry all i came down with covid so this got put on the back burner, i should be able to take a new look at this towards the end of the week

**Comment by chelnak** [comment]
@b4ldr No rush! Take your time and recover! 😄

**Pull request closed without merging**